File: src_py/542-01-matrix.py
"""
542. 01 Matrix
Difficulty: Medium

Given an m x n binary matrix mat, return the distance of the nearest 0 for
each cell. The distance between two adjacent cells is 1.
"""

from collections import deque
from typing import List


DIRECTIONS = [(0, 1), (0, -1), (1, 0), (-1, 0)]


def update_matrix_bfs(mat: List[List[int]]) -> List[List[int]]:
    rows, cols = len(mat), len(mat[0])
    dist = [[-1] * cols for _ in range(rows)]
    queue = deque()

    for r in range(rows):
        for c in range(cols):
            if mat[r][c] == 0:
                dist[r][c] = 0
                queue.append((r, c))

    while queue:
        r, c = queue.popleft()
        for dr, dc in DIRECTIONS:
            nr, nc = r + dr, c + dc
            if 0 <= nr < rows and 0 <= nc < cols and dist[nr][nc] == -1:
                dist[nr][nc] = dist[r][c] + 1
                queue.append((nr, nc))

    return dist


def update_matrix_dp(mat: List[List[int]]) -> List[List[int]]:
    rows, cols = len(mat), len(mat[0])
    INF = float("inf")
    dist = [[0 if mat[r][c] == 0 else INF for c in range(cols)] for r in range(rows)]

    # Pass 1: from top-left
    for r in range(rows):
        for c in range(cols):
            if r > 0:
                dist[r][c] = min(dist[r][c], dist[r - 1][c] + 1)
            if c > 0:
                dist[r][c] = min(dist[r][c], dist[r][c - 1] + 1)

    # Pass 2: from bottom-right
    for r in range(rows - 1, -1, -1):
        for c in range(cols - 1, -1, -1):
            if r < rows - 1:
                dist[r][c] = min(dist[r][c], dist[r + 1][c] + 1)
            if c < cols - 1:
                dist[r][c] = min(dist[r][c], dist[r][c + 1] + 1)

    return dist


if __name__ == "__main__":
    mat = [[0, 0, 0], [0, 1, 0], [1, 1, 1]]
    expected = [[0, 0, 0], [0, 1, 0], [1, 2, 1]]
    assert update_matrix_bfs(mat) == expected
    assert update_matrix_dp(mat) == expected

    mat2 = [[0, 0, 0], [0, 1, 0], [0, 0, 0]]
    assert update_matrix_bfs(mat2) == [[0, 0, 0], [0, 1, 0], [0, 0, 0]]
    assert update_matrix_dp(mat2) == [[0, 0, 0], [0, 1, 0], [0, 0, 0]]
    print("542 OK:", update_matrix_bfs(mat))

File: src/542-01-matrix.ts
import { check } from "./utils/check"

// solution 1 BFS
function updateMatrixBFS(mat: number[][]): number[][] {
  const rows = mat.length;
  const cols = mat[0].length;

  const dist: number[][] = Array.from({ length: rows }, () =>
    Array(cols).fill(Infinity)
  );

  const queue: [number, number][] = [];

  // 1. Add all zero cells as BFS starting points
  for (let r = 0; r < rows; r++) {
    for (let c = 0; c < cols; c++) {
      if (mat[r][c] === 0) {
        dist[r][c] = 0;
        queue.push([r, c]);
      }
    }
  }

  const directions = [
    [1, 0],
    [-1, 0],
    [0, 1],
    [0, -1],
  ];

  let head = 0;

  // 2. BFS from all zeros
  while (head < queue.length) {
    const [r, c] = queue[head++];

    for (const [dr, dc] of directions) {
      const nr = r + dr;
      const nc = c + dc;

      if (nr < 0 || nr >= rows || nc < 0 || nc >= cols) {
        continue;
      }

      // If going through current cell gives a shorter distance
      if (dist[nr][nc] > dist[r][c] + 1) {
        dist[nr][nc] = dist[r][c] + 1;
        queue.push([nr, nc]);
      }
    }
  }

  return dist;
}

function updateMatrixDP(mat: number[][]): number[][] {
  const rows = mat.length;
  const cols = mat[0].length;

  const dist: number[][] = Array.from({ length: rows }, () =>
    Array(cols).fill(Infinity)
  );

  // Initialize zeros
  for (let r = 0; r < rows; r++) {
    for (let c = 0; c < cols; c++) {
      if (mat[r][c] === 0) {
        dist[r][c] = 0;
      }
    }
  }

  // Pass 1: top-left to bottom-right
  // Check top and left
  for (let r = 0; r < rows; r++) {
    for (let c = 0; c < cols; c++) {
      if (r > 0) {
        dist[r][c] = Math.min(dist[r][c], dist[r - 1][c] + 1);
      }

      if (c > 0) {
        dist[r][c] = Math.min(dist[r][c], dist[r][c - 1] + 1);
      }
    }
  }

  // Pass 2: bottom-right to top-left
  // Check bottom and right
  for (let r = rows - 1; r >= 0; r--) {
    for (let c = cols - 1; c >= 0; c--) {
      if (r < rows - 1) {
        dist[r][c] = Math.min(dist[r][c], dist[r + 1][c] + 1);
      }

      if (c < cols - 1) {
        dist[r][c] = Math.min(dist[r][c], dist[r][c + 1] + 1);
      }
    }
  }

  return dist;
}

// ---- tests ----
{

  type Case = { name: string, mat: number[][], expected: number[][] }
  const cases: Case[] = [
    {
      name: "case1 single 1 in center",
      mat: [[0, 0, 0], [0, 1, 0], [0, 0, 0]],
      expected: [[0, 0, 0], [0, 1, 0], [0, 0, 0]],
    },
    {
      name: "case2 bottom row of 1s",
      mat: [[0, 0, 0], [0, 1, 0], [1, 1, 1]],
      expected: [[0, 0, 0], [0, 1, 0], [1, 2, 1]],
    },
    {
      name: "case3 single zero corner",
      mat: [[0, 1, 1], [1, 1, 1]],
      expected: [[0, 1, 2], [1, 2, 3]],
    },
  ]

  for (const c of cases) {
    check(`${c.name} (BFS)`, updateMatrixBFS(c.mat), c.expected)
    check(`${c.name} (DP)`, updateMatrixDP(c.mat), c.expected)
  }
}

// make this file a module so its declarations stay file-scoped
export {}
